File: verify/report.mjs
/**
 * Recompute the classification report from the confusion matrix in JavaScript.
 *
 * Same check as the other verifiers: derives precision, recall, F1, support
 * and macro-F1 from the raw 7x7 confusion matrix in ai/reports/metrics.json
 * and compares against the published report.
 *
 * Run: node verify/report.mjs [repo root]
 */
import { readFileSync } from "fs";
import { join, resolve } from "path";

const root = resolve(process.argv[2] || ".");
const d = JSON.parse(readFileSync(join(root, "ai", "reports", "metrics.json"), "utf8"));

const classes = d.classes;
const cm = d.test_confusion_matrix;
const report = d.test_classification_report;
const publishedF1 = d.test_macro_f1;
const n = classes.length;
const TOL = 5e-4;
let bad = 0;

function fail(msg) {
  console.log(`  FAIL: ${msg}`);
  bad++;
}

const f1s = [];
for (let i = 0; i < n; i++) {
  const tp = cm[i][i];
  let fp = 0;
  let fn = 0;
  for (let j = 0; j < n; j++) {
    if (j !== i) {
      fp += cm[j][i];
      fn += cm[i][j];
    }
  }
  const precision = tp + fp > 0 ? tp / (tp + fp) : 0;
  const recall = tp + fn > 0 ? tp / (tp + fn) : 0;
  const f1 = precision + recall > 0 ? (2 * precision * recall) / (precision + recall) : 0;
  f1s.push(f1);

  const name = classes[i];
  const pub = report[name] || {};
  for (const [metric, got] of [["precision", precision], ["recall", recall], ["f1-score", f1]]) {
    const pubVal = pub[metric];
    if (pubVal != null && Math.abs(got - pubVal) > TOL) {
      fail(`${name} ${metric}: published ${pubVal.toFixed(4)}, recomputed ${got.toFixed(4)}`);
    }
  }
}

const macroF1 = f1s.reduce((a, b) => a + b, 0) / n;
if (Math.abs(macroF1 - publishedF1) > TOL) {
  fail(`macro-F1: published ${publishedF1.toFixed(6)}, recomputed ${macroF1.toFixed(6)}`);
}

if (bad) {
  console.log(`JavaScript: ${bad} problem(s)`);
  process.exit(1);
}
console.log(`JavaScript: ${n} classes, macro-F1 ${macroF1.toFixed(4)} reproduced from confusion matrix`);

File: verify/report.py
"""Recompute the classification report from the confusion matrix in Python.

Same check as verify/report.c and verify/report.sql: derives precision, recall,
F1, support and macro-F1 from the raw 7x7 confusion matrix in
ai/reports/metrics.json and compares against the published report.

Shares no code with the training scripts.

Run: python3 verify/report.py <repo root>
"""
import json
import sys
from pathlib import Path

root = Path(sys.argv[1]) if len(sys.argv) > 1 else Path(".")
d = json.loads((root / "ai" / "reports" / "metrics.json").read_text())

classes = d["classes"]
cm = d["test_confusion_matrix"]
report = d["test_classification_report"]
published_f1 = d["test_macro_f1"]
n = len(classes)
bad = 0
TOL = 5e-4

def fail(msg):
    global bad
    print(f"  FAIL: {msg}")
    bad += 1

f1s = []
for i in range(n):
    tp = cm[i][i]
    fp = sum(cm[j][i] for j in range(n) if j != i)
    fn = sum(cm[i][j] for j in range(n) if j != i)
    support = sum(cm[i])
    precision = tp / (tp + fp) if (tp + fp) > 0 else 0.0
    recall = tp / (tp + fn) if (tp + fn) > 0 else 0.0
    f1 = 2 * precision * recall / (precision + recall) if (precision + recall) > 0 else 0.0
    f1s.append(f1)

    name = classes[i]
    pub = report.get(name, {})
    for metric, got in [("precision", precision), ("recall", recall), ("f1-score", f1)]:
        pub_val = pub.get(metric)
        if pub_val is not None and abs(got - pub_val) > TOL:
            fail(f"{name} {metric}: published {pub_val:.4f}, recomputed {got:.4f}")

macro_f1 = sum(f1s) / n
if abs(macro_f1 - published_f1) > TOL:
    fail(f"macro-F1: published {published_f1:.6f}, recomputed {macro_f1:.6f}")

if bad:
    print(f"Python: {bad} problem(s)")
    sys.exit(1)
print(f"Python: {n} classes, macro-F1 {macro_f1:.4f} reproduced from confusion matrix")
